src/cmd/services/m3coordinator/ingest/carbon/ingest.go

```diff
@@ -148,9 +149,107 @@ type ingester struct {
 	metrics              carbonIngesterMetrics
 	tagOpts              models.TagOptions
 
+	lineResourcesPool pool.ObjectPool
+
+	sync.RWMutex
 	rules []ruleAndRegex
+}
 
-	lineResourcesPool pool.ObjectPool
+func (i *ingester) OnUpdate(clusterNamespaces m3.ClusterNamespaces) {
+	rules := i.regenerateIngestionRules(clusterNamespaces)
+	if rules == nil {
+		namespaces := make([]string, 0, len(clusterNamespaces))
+		for _, ns := range clusterNamespaces {
+			namespaces = append(namespaces, ns.NamespaceID().String())
+		}
+		i.logger.Warn("generated empty carbon ingestion rules from latest cluster namespaces update. leaving"+
+			" current set of rules as-is.", zap.Strings("namespaces", namespaces))
+		return
+	}
+
+	compiledRules, err := i.compileRules(*rules)
+	if err != nil {
+		i.logger.Error("failed to compile latest rules. continuing to use existing carbon ingestion "+
+			"rules", zap.Error(err))
+		return
+	}
+
+	i.Lock()
+	i.rules = compiledRules
+	i.Unlock()
+}
+
+func (i *ingester) regenerateIngestionRules(clusterNamespaces m3.ClusterNamespaces) *CarbonIngesterRules {
+	var (
+		rules = &CarbonIngesterRules{
+			Rules: i.opts.IngesterConfig.RulesOrDefault(clusterNamespaces),
+		}
+		namespacesByRetention = make(map[m3.RetentionResolution]m3.ClusterNamespace, len(clusterNamespaces))
+	)
+
+	for _, ns := range clusterNamespaces {
+		if ns.Options().Attributes().MetricsType == storagemetadata.AggregatedMetricsType {
+			namespacesByRetention[m3.RetentionResolution{
+				Resolution: ns.Options().Attributes().Resolution,
+				Retention:  ns.Options().Attributes().Retention,
+			}] = ns
+		}
+	}
+
+	// Validate rule policies.
+	for _, rule := range rules.Rules {
+		// Sort so we can detect duplicates.
+		sort.Slice(rule.Policies, func(i, j int) bool {
+			if rule.Policies[i].Resolution == rule.Policies[j].Resolution {
+				return rule.Policies[i].Retention < rule.Policies[j].Retention
+			}
+
+			return rule.Policies[i].Resolution < rule.Policies[j].Resolution
+		})
+
+		var lastPolicy config.CarbonIngesterStoragePolicyConfiguration
+		for idx, policy := range rule.Policies {
+			if idx > 0 && policy == lastPolicy {
```

Comment:
nit; don't think it's strictly necessary to check `idx > 0` unless `policy` can be `nil`

Reply:
Ah, I bet this was to prevent the last policy check the first time through, but no policy should match the default policy of 0 resolution and 0 retention, so I think you're right.